x/programs/rust/wasmlanche_sdk/src/types/mod.rs

```diff
@@ -1,73 +1,142 @@
+        // Find the first null byte and only print up to that point.
+        let null_pos = self.0.iter().position(|&b| b == b'\0').unwrap_or(Self::LEN);
+        String::from_utf8_lossy(&self.0[..null_pos])
+            .to_string()
+            .fmt(f)
     }
 }
 
-impl From<Address> for ProgramValue {
-    fn from(value: Address) -> Self {
-        ProgramValue::AddressObject(value)
+impl From<String> for Bytes32 {
+    fn from(value: String) -> Self {
+        let mut bytes: [u8; Self::LEN] = [0; Self::LEN];
+        bytes[..value.len()].copy_from_slice(value.as_bytes());
+        Self(bytes)
     }
 }
 
-impl From<i64> for Address {
+impl From<i64> for Bytes32 {
     fn from(value: i64) -> Self {
         let bytes: [u8; Self::LEN] = unsafe {
             // We want to copy the bytes here, since [value] represents a ptr created by the host
             std::slice::from_raw_parts(value as *const u8, Self::LEN)
                 .try_into()
                 .unwrap()
         };
-        Self { bytes }
+        Self(bytes)
+    }
+}
+
+pub trait Argument {
+    fn as_bytes(&self) -> Cow<'_, [u8]>;
+    fn is_primitive(&self) -> bool;
+    fn len(&self) -> usize {
+        self.as_bytes().len()
+    }
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+impl Argument for Bytes32 {
+    fn as_bytes(&self) -> Cow<'_, [u8]> {
+        Cow::Borrowed(&self.0)
+    }
+    fn is_primitive(&self) -> bool {
+        false
     }
 }
 
-impl From<ProgramValue> for i64 {
-    fn from(value: ProgramValue) -> Self {
-        match value {
-            ProgramValue::IntObject(i) => i,
-            _ => panic!("Cannot conver to i64"),
-        }
+impl Argument for Address {
+    fn as_bytes(&self) -> Cow<'_, [u8]> {
+        Cow::Borrowed(self.0.as_bytes())
+    }
+    fn is_primitive(&self) -> bool {
+        false
+    }
+}
+
+impl Argument for i64 {
+    fn as_bytes(&self) -> Cow<'_, [u8]> {
+        Cow::Owned(self.to_be_bytes().to_vec())
+    }
+    fn is_primitive(&self) -> bool {
+        true
+    }
+}
+
+impl Argument for Context {
+    fn as_bytes(&self) -> Cow<'_, [u8]> {
+        self.program_id.as_bytes()
+    }
+    fn is_primitive(&self) -> bool {
+        true
+    }
+}
+
+// Implement Box<dyn Argument> for all types that implement Argument.
+impl From<i64> for Box<dyn Argument> {
+    fn from(value: i64) -> Self {
+        Box::new(value)
+    }
+}
+
+impl From<Bytes32> for Box<dyn Argument> {
+    fn from(value: Bytes32) -> Self {
+        Box::new(value)
+    }
+}
+
+impl From<Address> for Box<dyn Argument> {
+    fn from(value: Address) -> Self {
+        Box::new(value)
     }
 }
 
-impl From<ProgramValue> for ProgramContext {
-    fn from(value: ProgramValue) -> Self {
-        match value {
-            ProgramValue::ProgramObject(i) => i,
-            _ => panic!("Cannot conver to ProgramContext"),
-        }
+impl From<Context> for Box<dyn Argument> {
+    fn from(value: Context) -> Self {
+        Box::new(value)
```

Comment:
All of these are unnecessary. The compiler automatically infers the type when necessary.
```rust
let x: Box<dyn Argument> = Box::new(0i64);
```
The above will just work.